fvm/src/kernel/default.rs

```diff
@@ -1006,41 +1007,102 @@ impl<C> EventOps for DefaultKernel<C>
+
         if self.read_only {
             return Err(syscall_error!(ReadOnly; "cannot emit events while read-only").into());
         }
-        let len = raw_evt.len();
-        let t = self
-            .call_manager
-            .charge_gas(self.call_manager.price_list().on_actor_event_validate(len))?;
 
-        // This is an over-estimation of the maximum event size, for safety. No valid event can even
-        // get close to this. We check this first so we don't try to decode a large event.
-        const MAX_ENCODED_SIZE: usize = 1 << 20;
-        if raw_evt.len() > MAX_ENCODED_SIZE {
-            return Err(syscall_error!(IllegalArgument; "event WAY too large").into());
+        if raw_val.len() > MAX_VALUE_LEN {
+            return Err(
+                syscall_error!(IllegalArgument; "event total values exceeded max size: {} > {MAX_VALUE_LEN}", raw_val.len()).into(),
+            );
         }
 
-        let actor_evt = {
-            let res = match panic::catch_unwind(|| {
-                fvm_ipld_encoding::from_slice(raw_evt).or_error(ErrorNumber::IllegalArgument)
-            }) {
-                Ok(v) => v,
-                Err(e) => {
-                    log::error!("panic when decoding event cbor from actor: {:?}", e);
-                    Err(syscall_error!(IllegalArgument; "panic when decoding event cbor from actor").into())
-                }
+        let total_buf_len = raw_evt.len() + raw_key.len() + raw_val.len();
+
+        let _t = self.call_manager.charge_gas(
+            self.call_manager
+                .price_list()
+                .on_actor_event_validate(total_buf_len),
+        )?;
+
+        const BYTES_PER_ENTRY: usize = 24;
+        if raw_evt.len() % BYTES_PER_ENTRY != 0 {
+            return Err(syscall_error!(IllegalArgument; "event not divisible by 24 bytes").into());
+        }
+        let nr_entries = raw_evt.len() / BYTES_PER_ENTRY;
+
+        if nr_entries > MAX_NR_ENTRIES {
+            return Err(syscall_error!(IllegalArgument; "event exceeded max entries: {} > {MAX_NR_ENTRIES}", nr_entries).into());
+        }
+
+        let mut key_offset: usize = 0;
+        let mut val_offset: usize = 0;
+
+        let mut entries: Vec<Entry> = Vec::with_capacity(nr_entries);
+        for i in 0..nr_entries {
+            let offset = i * BYTES_PER_ENTRY;
+            let view = &raw_evt[offset..offset + BYTES_PER_ENTRY];
+
+            // parse the fixed sized fields from the raw_evt buffer
+            //
+            let flags = match Flags::from_bits(BigEndian::read_u64(&view[..8])) {
+                Some(f) => f,
+                None => return Err(syscall_error!(IllegalArgument, "invalid flag").into()),
             };
-            t.stop();
-            res
-        }?;
-        validate_actor_event(&actor_evt)?;
+            let key_len = BigEndian::read_u32(&view[8..12]);
+            let codec = BigEndian::read_u64(&view[12..20]);
+            let val_len = BigEndian::read_u32(&view[20..24]);
```

Comment:
Given that we use little-endian for return values, we should likely do the same here. I.e., we can literally transmute the input array into an array of packed structs.

Reply:
```rust

// define this in shared
#[repr(C, packed)]
struct EventMeta {
    flags: Flags, // repr ut64
    codec: u64, // order matters for packing.
    key_len: u32,
    value_len: u32,
} 
```

Although... we'll likely need to transmute one-by-one and copy as the input slice _may_ not be aligned.

Reply:
This is less critical on the FVM side, but it'll save time and avoid swapping around bytes in wasm.

Reply:
Note: `Flags` will have to be decorated with `#[repr(transparent)]` for this to work.